### DISCUSSION TITLE
Can Alltalk_tts support Indonesian language?

### DISCUSSION THREAD

**Discussion by SyamsQ** [erew123/alltalk_tts#460] [Q&A]
Can Alltalk_tts support Indonesian language?

**Comment by erew123** [comment]
I can support whatever language the underlying TTS engine supports https://github.com/erew123/alltalk_tts/wiki/AllTalk-V2-QuickStart-Guide#currently-available-tts-engines 

Please see the manufactures/developers websites for what their TTS engine's are capable of.

Thanks

**Reply by SyamsQ** [reply]
How do I add a language to Alltalk? I see some of the TTS engines you shared support Indonesian.

**Reply by erew123** [reply]
Hi @SyamsQ Which engine do you believe supports Indonesian? I dont see any of them listing Indonesian

**Reply by SyamsQ** [reply]
The ones I found included **Vits** and **F5-TTS**, supporting Indonesian and Malaysian as adjacent languages.

I've even tried HF Vits Space in Indonesian.

As for **Xtts Coqui**, based on info on Github, it supports 1100 languages.

Please guide and help.
![image](https://github.com/user-attachments/assets/cd2098ff-baa6-4965-85a4-a87b08276e3d)
![image](https://github.com/user-attachments/assets/40c327ff-5920-4807-b448-59f14fe52dca)
![image](https://github.com/user-attachments/assets/3bc466e1-1d98-4bc5-a724-771e89b2585a)
![image](https://github.com/user-attachments/assets/20cc3065-1e6f-4b5b-8842-c51eb012e8fa)
![image](https://github.com/user-attachments/assets/6387a2d3-2062-404d-a800-5309669e4d9d)
![image](https://github.com/user-attachments/assets/b0f76792-3531-4125-8cad-817bb6e033ec)
![image](https://github.com/user-attachments/assets/73e35228-3889-41c7-a6da-d1969abc8fec)
![image](https://github.com/user-attachments/assets/5e7ab3af-26a6-4fe4-baf2-718f2aa1c07e)
![image](https://github.com/user-attachments/assets/c93b48ea-05cc-4945-82b3-8ffc038ba067)

**Comment by erew123** [comment]
### Generally adding a language specific option to AT

To add a language option to AllTalk's interface and API you can edit script.py and tts_server.py to add:

`ms` for Malaysian/Malay
`id` for Indonesian

https://github.com/erew123/alltalk_tts/blob/alltalkbeta/script.py#L4290-L4313

https://github.com/erew123/alltalk_tts/blob/alltalkbeta/tts_server.py#L1627

### VITS models

VITS models, the language support is built into the actual model, so you cannot select/specify the language, it just is whatever the actual model is. So if you download a Indonesian model, put it in a folder, TTS will speak Indonesian (is my understanding). Hence adding `ms` or `id` language options to AT would have no bearing/impact/use.

![image](https://github.com/user-attachments/assets/e18850c6-832b-47c9-8a3c-e2a3373c0576)

At minimum a VITS model needs to be in its own folder (as shown above) and have a `model_file.pth` or `model.pth` in the folder, along with a `config.json` file. The code that scans for models can be seen here.

https://github.com/erew123/alltalk_tts/blob/alltalkbeta/system/tts_engines/vits/model_engine.py#L267-L275

### F5 models
As far as I am aware with F5 models, the vocab.txt handles the tokenization of the language. I assume there will be instructions provided on the peoples sites for their models. If certain files are not available from them, you can try the files that come with the default downloads of an F5 model (as in download one of the standard models then merge in what files you need).

As for how AllTalk looks for F5 models, the models should be placed in the following structure:

```
models/
└── f5tts/
    └── *tts_v*/               # Folder name must contain "tts_v" within its name
        ├── model_*.safetensors # Primary model file (or .pt as fallback)
        ├── vocab.txt          # Required vocabulary file
        └── vocos/             # Required vocos folder
            ├── config.yaml    # Required vocos config
            └── pytorch_model.bin # Required vocos model
```

Key requirements:
1. Main folder must be `models/f5tts/`
2. Each model folder name must contain `tts_v` (e.g., `f5tts_v1`, `mytts_v2`, etc.)
3. Each model folder must contain:
   - A model file (either `.safetensors` or `.pt` format)
   - A `vocab.txt` file
   - A `vocos` subfolder containing:
     - `config.yaml`
     - `pytorch_model.bin`

The code will first look for:
1. `model_*.safetensors` files
2. Any `.safetensors` files
3. `model_*.pt` files
4. Any `.pt` files

Be aware that pt files may contain viruses and are not recommended.

If any required files are missing, the model will be considered invalid and won't be loaded.

You will have to follow whatever instructions those 3rd parties give for setting up their model.

**Reply by SyamsQ** [reply]
There is a requirement to run this command, how to run it in All_talk? Does it have to be done?
![image](https://github.com/user-attachments/assets/336c73ab-24d5-45f8-a7ef-893ad61da9c7)
